File: 116.populating-next-right-pointers-in-each-node.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if (!root) {
            return;
        }
        connect(root->left);
        connect(root->right);
        TreeLinkNode *left = root->left, *right = root->right;
        while (left && right) {
            left->next = right;
            left = left->right;
            right = right->left;
        }
    }
};

File: 116.populating-next-right-pointers-in-each-node.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if not root:
            return;
        self.connect(root.left)
        self.connect(root.right)
        left = root.left
        right = root.right
        while left and right:
            left.next = right
            left = left.right
            right = right.left

